File: apps/ios/Sources/Gateway/CloudflareAccessClient.swift
import Foundation

struct CloudflareAccessClient: Sendable {
    typealias Request = @Sendable (URLRequest, Int) async throws -> (Data, HTTPURLResponse)

    let request: Request

    init(request: @escaping Request = Self.send) {
        self.request = request
    }

    /// Check ordinary ingress first: the metadata endpoint returns 200 even before authentication.
    func discover(gatewayURL: URL, session: CloudflareAccessSession? = nil) async throws
        -> CloudflareAccessApplication?
    {
        let origin = try CloudflareAccessOrigin(gatewayURL)
        guard var components = URLComponents(url: gatewayURL, resolvingAgainstBaseURL: false) else {
            throw CloudflareAccessError.invalidGateway
        }
        components.scheme = "https"
        if components.path.isEmpty {
            components.path = "/"
        }
        guard let url = components.url else { throw CloudflareAccessError.invalidGateway }
        var probe = URLRequest(url: url)
        probe.setValue(session?.authorizationHeader(for: url), forHTTPHeaderField: "Cf-Access-Token")
        let (_, response) = try await self.request(probe, 0)
        guard Self.isChallenge(response, origin: origin) else { return nil }

        var metadataRequest = URLRequest(url: url)
        metadataRequest.httpMethod = "HEAD"
        metadataRequest.setValue("true", forHTTPHeaderField: "Cf-Access-Metadata-Request")
        metadataRequest.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        let (_, metadataResponse) = try await self.request(metadataRequest, 0)
        guard metadataResponse.statusCode == 200,
              let token = metadataResponse.value(forHTTPHeaderField: "Cf-Access-Metadata")
        else {
            throw CloudflareAccessError.invalidApplication
        }
        do {
            // Only a constrained Cloudflare issuer can supply the key. The signature then
            // authenticates the hostname and audience before any browser or credential use.
            let metadata = try CloudflareAccessJWT.decode(CloudflareAccessJWT.Metadata.self, token: token)
            let application = try CloudflareAccessJWT.application(metadata: metadata, origin: origin)
            try await CloudflareAccessJWT.verify(token, jwks: self.keys(for: application))
            return application
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw CloudflareAccessError.invalidApplication
        }
    }

    static func isChallenge(_ response: HTTPURLResponse, origin: CloudflareAccessOrigin) -> Bool {
        guard let responseURL = response.url, origin.contains(responseURL) else { return false }
        // cloudflared also recognizes the decoded login path on a 302. The redirect is
        // only a hint: discovery still verifies signed metadata from the original URL.
        if response.statusCode == 302,
           let location = response.value(forHTTPHeaderField: "Location"), !location.isEmpty,
           var target = URLComponents(string: location, encodingInvalidCharacters: false)
        {
            // Go keeps scheme-less triple-leading slashes in the path; Foundation parses an empty authority.
            // Restore that path on the original authority for local normalization only, never for a request.
            if location.hasPrefix("///") {
                target.percentEncodedPath = "//" + target.percentEncodedPath
                target.host = responseURL.host
                target.port = responseURL.port
            }
            if target.url(relativeTo: responseURL)?.absoluteURL.standardized.path
                .hasPrefix("/cdn-cgi/access/login") == true
            {
                return true
            }
        }
        guard [301, 302, 303, 307, 308, 401, 403].contains(response.statusCode),
              let header = response.value(forHTTPHeaderField: "WWW-Authenticate"), header.utf8.count <= 8192
        else { return false }
        let parts = header.split(maxSplits: 1, whereSeparator: { $0.isWhitespace })
        guard parts.count == 2, ["cloudflare-access", "bearer"].contains(parts[0].lowercased()),
              let expression = try? NSRegularExpression(pattern: #"(?:^|[,\s])resource_metadata\s*=\s*"([^"]+)""#),
              let match = expression.firstMatch(
                  in: String(parts[1]),
                  range: NSRange(parts[1].startIndex..., in: parts[1])),
              let range = Range(match.range(at: 1), in: parts[1]),
              let metadataURL = URL(string: String(parts[1][range])),
              metadataURL.scheme?.lowercased() == "https", origin.contains(metadataURL), metadataURL.query == nil
        else { return false }
        // RFC 9728 puts resource paths after the namespace. This is only a
        // challenge hint; signed metadata is still requested from the original URL.
        let namespace = "/.well-known/cloudflare-access-protected-resource"
        return metadataURL.path == namespace || metadataURL.path.hasPrefix(namespace + "/")
    }

    func verifiedSession(token: String, application: CloudflareAccessApplication) async throws
        -> CloudflareAccessSession
    {
        try await CloudflareAccessJWT.verify(token, jwks: self.keys(for: application))
        let claims = try CloudflareAccessJWT.appClaims(token, application: application)
        var identityRequest = URLRequest(url: application.origin.url
            .appendingPathComponent("cdn-cgi/access/get-identity"))
        // Access's identity endpoint consumes its session cookie. This one request
        // has no cookie jar or redirects; native Gateway traffic uses Cf-Access-Token.
        identityRequest.setValue("CF_Authorization=\(token)", forHTTPHeaderField: "Cookie")
        let (data, response) = try await self.request(identityRequest, Self.maximumResponseBytes)
        struct Identity: Decodable {
            let userUUID: String

            enum CodingKeys: String, CodingKey {
                case userUUID = "user_uuid"
            }
        }
        guard response.statusCode == 200,
              let identity = try? JSONDecoder().decode(Identity.self, from: data),
              identity.userUUID == claims.sub
        else { throw CloudflareAccessError.invalidSession }
        try Task.checkCancellation()
        return CloudflareAccessSession(
            application: application,
            subject: claims.sub,
            token: token,
            expiresAt: Date(timeIntervalSince1970: claims.exp))
    }

    private func keys(for application: CloudflareAccessApplication) async throws -> Data {
        guard let host = application.issuer.host,
              try CloudflareAccessJWT.issuer(authDomain: host) == application.issuer
        else { throw CloudflareAccessError.invalidApplication }
        let request = URLRequest(url: application.issuer.appendingPathComponent("cdn-cgi/access/certs"))
        let (data, response) = try await self.request(request, Self.maximumResponseBytes)
        guard response.statusCode == 200 else { throw CloudflareAccessError.invalidApplication }
        return data
    }

    private final class NoRedirects: NSObject, URLSessionTaskDelegate {
        func urlSession(
            _ session: URLSession,
            task: URLSessionTask,
            willPerformHTTPRedirection response: HTTPURLResponse,
            newRequest request: URLRequest,
            completionHandler: @escaping (URLRequest?) -> Void)
        {
            completionHandler(nil)
        }
    }

    static let userAgent = "OpenClaw CloudflareAccess (cloudflared/2026.8.3)"
    static let maximumResponseBytes = 1_048_576

    static func send(_ request: URLRequest, maximumBytes: Int) async throws -> (Data, HTTPURLResponse) {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpCookieStorage = nil
        configuration.urlCredentialStorage = nil
        configuration.urlCache = nil
        configuration.httpShouldSetCookies = false
        configuration.timeoutIntervalForRequest = 15
        configuration.timeoutIntervalForResource = max(30, request.timeoutInterval)
        let session = URLSession(configuration: configuration, delegate: NoRedirects(), delegateQueue: nil)
        defer { session.invalidateAndCancel() }
        do {
            let (bytes, response) = try await session.bytes(for: request)
            guard let response = response as? HTTPURLResponse, response.url == request.url else {
                throw CloudflareAccessError.connectionFailed
            }
            var data = Data()
            if maximumBytes > 0 {
                for try await byte in bytes {
                    guard data.count < maximumBytes else { throw CloudflareAccessError.connectionFailed }
                    data.append(byte)
                }
            }
            return (data, response)
        } catch {
            if Task.isCancelled { throw CancellationError() }
            throw CloudflareAccessError.connectionFailed
        }
    }
}

File: apps/android/app/src/main/java/ai/openclaw/app/gateway/CloudflareAccessClient.kt
package ai.openclaw.app.gateway

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import okhttp3.Call
import okhttp3.Callback
import okhttp3.CookieJar
import okhttp3.Headers
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

internal class CloudflareAccessClient(
  private val request: suspend (Request, Int, Long) -> Reply = ::send,
) {
  class Reply(
    val url: String,
    val code: Int,
    val headers: Headers,
    val body: ByteArray,
  )

  /** A metadata 200 proves protection, not that the current WARP/manual-header session needs login. */
  suspend fun discover(
    gatewayUrl: String,
    session: CloudflareAccessSession? = null,
    customHeaders: Map<String, String> = emptyMap(),
  ): CloudflareAccessApplication? {
    val origin = CloudflareAccessOrigin.from(gatewayUrl)
    val url = gatewayUrl.replaceFirst(Regex("^wss:", RegexOption.IGNORE_CASE), "https:")
    val probe = Request.Builder().url(url)
    GatewayCustomHeaders.sanitized(customHeaders).forEach { (name, value) -> probe.header(name, value) }
    session?.authorizationHeader(url)?.let { probe.header("Cf-Access-Token", it) }
    val response = request(probe.build(), 0, 15)
    if (!isChallenge(response, origin)) return null
    try {
      val metadata =
        request(
          Request
            .Builder()
            .url(url)
            .head()
            .header("Cf-Access-Metadata-Request", "true")
            .header("User-Agent", userAgent)
            .build(),
          0,
          15,
        )
      val token = metadata.headers["Cf-Access-Metadata"]
      if (metadata.code != 200 || token == null) throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidApplication)
      val application = CloudflareAccessJWT.application(token, origin)
      CloudflareAccessJWT.verify(token, keys(application))
      return application
    } catch (error: CancellationException) {
      throw error
    } catch (error: IOException) {
      // Transport failures retain TLS diagnostics; they do not invalidate signed metadata.
      throw error
    } catch (_: Exception) {
      throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidApplication)
    }
  }

  suspend fun verifiedSession(
    token: String,
    application: CloudflareAccessApplication,
  ): CloudflareAccessSession {
    CloudflareAccessJWT.verify(token, keys(application))
    val claims = CloudflareAccessJWT.appClaims(token, application)
    // get-identity consumes the Access cookie. Only this exact no-redirect request receives it;
    // native Gateway traffic projects Cf-Access-Token instead and never uses a cookie jar.
    val identity =
      request(
        Request
          .Builder()
          .url(
            application.origin.uri
              .resolve("/cdn-cgi/access/get-identity")
              .toString(),
          ).header("Cookie", "CF_Authorization=$token")
          .build(),
        maximumResponseBytes,
        15,
      )
    val subject =
      runCatching {
        (Json.parseToJsonElement(identity.body.decodeToString()).jsonObject["user_uuid"] as? JsonPrimitive)
          ?.takeIf(JsonPrimitive::isString)
          ?.content
      }.getOrNull()
    if (identity.code != 200 || subject != claims.subject) throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidSession)
    kotlin.coroutines.coroutineContext.ensureActive()
    return CloudflareAccessSession(application, claims.subject, claims.expiresAt, token)
  }

  private suspend fun keys(application: CloudflareAccessApplication): ByteArray {
    if (CloudflareAccessJWT.issuer(application.issuer.host.orEmpty()) != application.issuer) {
      throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidApplication)
    }
    val response = request(Request.Builder().url(application.issuer.resolve("/cdn-cgi/access/certs").toString()).build(), maximumResponseBytes, 15)
    if (response.code != 200) throw CloudflareAccessException(CloudflareAccessException.Kind.InvalidApplication)
    return response.body
  }

  companion object {
    const val userAgent = "OpenClaw CloudflareAccess (cloudflared/2026.8.3)"
    const val maximumResponseBytes = 1_048_576
    private val client =
      OkHttpClient
        .Builder()
        .followRedirects(false)
        .followSslRedirects(false)
        .cookieJar(CookieJar.NO_COOKIES)
        .cache(null)
        .build()

    fun isChallenge(
      response: Reply,
      origin: CloudflareAccessOrigin,
    ): Boolean {
      if (!origin.contains(response.url)) return false
      // cloudflared treats a decoded login-path prefix on a 302 as a hint, never as application identity.
      // Metadata is still verified at the original gateway URL; the Location is never followed.
      if (response.code == 302 &&
        runCatching {
          val location = response.headers["Location"]?.takeIf { it.isNotEmpty() } ?: return@runCatching false
          val path = resolvedRedirectPath(java.net.URI(response.url), location)
          path?.startsWith("/cdn-cgi/access/login") == true
        }.getOrDefault(false)
      ) {
        return true
      }
      if (response.code !in setOf(301, 302, 303, 307, 308, 401, 403)) return false
      val header = response.headers["WWW-Authenticate"] ?: return false
      if (header.length > 8192) return false
      val parts = header.split(Regex("\\s+"), limit = 2)
      if (parts.size != 2 || parts[0].lowercase() !in setOf("cloudflare-access", "bearer")) return false
      val metadata = Regex("(?:^|[,\\s])resource_metadata\\s*=\\s*\"([^\"]+)\"").find(parts[1])?.groupValues?.get(1) ?: return false
      if (!origin.contains(metadata)) return false
      val uri = java.net.URI(metadata)
      // RFC 9728 appends resource paths to the namespace. This only admits the
      // challenge; signed metadata is still requested from the original URL.
      val namespace = "/.well-known/cloudflare-access-protected-resource"
      return uri.scheme.equals("https", ignoreCase = true) && uri.rawQuery == null &&
        (uri.path == namespace || uri.path.startsWith("$namespace/"))
    }

    private fun resolvedRedirectPath(
      base: java.net.URI,
      reference: String,
    ): String? {
      val uri = java.net.URI(reference)
      // Go preserves scheme-less triple-leading slashes as path; URI discards the empty authority.
      val raw = if (reference.startsWith("///")) reference.substringBefore('?').substringBefore('#') else uri.rawPath ?: return null
      val path =
        when {
          uri.isAbsolute || uri.rawAuthority != null || raw.startsWith("/") -> raw
          raw.isEmpty() -> base.rawPath
          else -> base.rawPath.substringBeforeLast('/', "") + "/" + raw
        }
      if (path.isEmpty()) return ""
      // Go resolves literal dots on the escaped path. URI.normalize also collapses
      // empty segments, while OkHttp resolves encoded dots; neither preserves this contract.
      val segments = ArrayDeque<String>()
      for (segment in path.split('/').drop(1)) {
        when (segment) {
          "." -> Unit
          ".." -> if (segments.isNotEmpty()) segments.removeLast()
          else -> segments.addLast(segment)
        }
      }
      if (path.endsWith("/.") || path.endsWith("/..")) segments.addLast("")
      // A relative dot prefix keeps a leading // in the path, never in the authority.
      return java.net
        .URI("./" + segments.joinToString("/"))
        .path
        .drop(1)
    }

    suspend fun send(
      request: Request,
      maximumBytes: Int,
      timeoutSeconds: Long,
    ): Reply =
      suspendCancellableCoroutine { continuation ->
        val call =
          client
            .newBuilder()
            .callTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .build()
            .newCall(request)
        continuation.invokeOnCancellation { call.cancel() }
        call.enqueue(
          object : Callback {
            override fun onFailure(
              call: Call,
              e: IOException,
            ) {
              if (continuation.isActive) continuation.resumeWithException(e)
            }

            override fun onResponse(
              call: Call,
              response: Response,
            ) {
              try {
                response.use {
                  if (response.request.url != request.url) throw CloudflareAccessException(CloudflareAccessException.Kind.ConnectionFailed)
                  val bytes =
                    if (maximumBytes == 0) {
                      byteArrayOf()
                    } else {
                      val source = response.body.source()
                      source.request(maximumBytes.toLong() + 1)
                      if (source.buffer.size > maximumBytes) throw CloudflareAccessException(CloudflareAccessException.Kind.ConnectionFailed)
                      source.readByteArray()
                    }
                  continuation.resume(Reply(response.request.url.toString(), response.code, response.headers, bytes))
                }
              } catch (error: IOException) {
                if (continuation.isActive) continuation.resumeWithException(error)
              } catch (_: Exception) {
                if (continuation.isActive) continuation.resumeWithException(CloudflareAccessException(CloudflareAccessException.Kind.ConnectionFailed))
              }
            }
          },
        )
      }
  }
}
